runtime/interpreter/value.go

```diff
@@ -15023,6 +15194,88 @@ func NewEnumCaseValue(
 	return v
 }
 
+func (v *CompositeValue) getBaseValue(interpreter *Interpreter, locationRange LocationRange) *EphemeralReferenceValue {
+	return v.base
+}
+
+func (v *CompositeValue) setBaseValue(interpreter *Interpreter, base *CompositeValue) {
+	attachmentType, ok := interpreter.MustConvertStaticToSemaType(v.StaticType(interpreter)).(*sema.CompositeType)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	// the super reference can only be borrowed with the declared type of the attachment's base
+	v.base = NewEphemeralReferenceValue(interpreter, false, base, attachmentType.GetBaseType())
+}
+
+func (v *CompositeValue) getAttachmentValue(interpreter *Interpreter, locationRange LocationRange, ty sema.Type) Value {
+	return v.GetMember(interpreter, locationRange, attachmentNamePrefix+ty.QualifiedString())
+}
+
+func (v *CompositeValue) forEachAttachment(interpreter *Interpreter, locationRange LocationRange, f func(*CompositeValue)) {
+	iterator, err := v.dictionary.Iterator()
+	if err != nil {
+		panic(errors.NewExternalError(err))
+	}
+
+	interpreter.SharedState.inAttachmentIteration[v] = struct{}{}
+	defer delete(interpreter.SharedState.inAttachmentIteration, v)
```

Comment:
Every time we enter an iteration over a composite's attachment we store that composite in this set, and remove it when we're done. This can't just be a `bool` like the storage iteration because we only want to restrict mutations of this specific composite's attachments, not all attachments generally.